[feature] Add SignerSet implementation

pfletcherhill opened the pull request:
Updated version of SignerSet implementation. Original work done by @drshrey. The SignerSet spec has changed a lot since this work was initially done, so here's a summary of what's included.

### What's included (high-level):
* SignerSet that stores custody addresses and lets those custody addresses add and remove delegate signers
* Scheme for revoking signers and dropping messages across all sets that were signed by the revoked signer
* Replace roots and signer changes with the SignerSet in the engine and across the codebase
* Remove simulator (it will be re-added in another repo in phase 3 of our hub launch plan)
* Update readme

### About SignerSet implementation
While this implementation is much simpler than prior SignerSet specs, there are still some quirks worth mentioning.
* Custody addresses are represented in `custodyAdds` and `custodyRemoves` and moved between the sets based on whether they are added or removed
* Custody addresses can only be added via Register and Transfer events from the Farcaster ID Registry (FIR), each of which has a block number, log index, and transaction hash
* When a transfer happens in the FIR, the old custody address is not immediately removed from the SignerSet, because that would immediately drop all messages signed by delegates of that address
* Custody addresses can only be removed by merging a `CustodyRemoveAll` message that is signed by a custody address. When accepted, the `CustodyRemoveAll` message removes all custody addresses that were added before the signer of the `CustodyRemoveAll` was added (based on the FIR events).
* Custody addresses function like a LWW set, where the order of FIR events is used to resolve conflicts. This setup means that custody addresses can be re-added after being removed, which is consistent with behavior in the FIR.
* Delegate signers can only be added via SignerAdd messages signed by custody addresses, and they can only be removed via SignerRemove messages signed by custody addresses.
* Unlike past designs, there is no message ordering, so a SignerRemove message can be accepted before a SignerAdd that references the same signer. This feature means the consistency guarantees (i.e. flexibility) of the SignerSet are equivalent to our other CRDTs.
* The signer sets also operate like a LWW set, using the message signer's FIR events to resolve conflicts.
* Like custody addresses, delegate signers can be re-added after they've been removed.

### About signer revocation
When a signer (custody addresses or delegates) is removed, all messages that were signed by that signer should be dropped, so the state matches what it would have been if the signer had been removed before the messages were accepted. Here are details about how we do signer revocation in this PR:
* The SignerSet now extends EventEmitter, so other parts of the codebase (i.e. the engine) can subscribe to SignerSet events. Eventually, all our CRDTs should be event emitters so devs can stream data changes from the CRDTs.
* The SignerSet has `addCustody`, `removeCustody`, `addSigner`, and `removeSigner` events
* Internally, right before a custody address is removed, the SignerSet drops all messages signed by that custody address
* The engine subscribes to events from the SignerSet, and when a signer is removed it calls a new `revokeSigner` method in all the CRDTs except the SignerSet
* Each `revokeSigner` takes the signer that was removed and drops all messages from its state that were signed by that signer

### SignerAdd and SignerRemove types
When signers were organized in a tree, it made sense to use "parent" and "child" terminology. But now that we only have custody addresses and delegate signers, I've updated the SignerAdd and SignerRemove bodies to be consistent:
```
type SignerAddBody = {
  delegate: string;
  edgeHash: string;
  delegateSignature: string;
  delegateSignatureType: SignatureAlgorithm.Ed25519;
  schema: 'farcaster.xyz/schemas/v1/signer-add';
};

type SignerRemoveBody = {
  delegate: string;
  schema: 'farcaster.xyz/schemas/v1/signer-remove';
};
```
The SignerEdge type also changes due to this naming:
```
type SignerEdge = {
  custody: string;
  delegate: string;
};
```

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    // If parent exists in vAdds or custody
+    if (this._vertexAdds.has(parentPubKey) || this._custodySigners.has(parentPubKey)) {
+      // If child does not exist in vertices
+      if (!this._vertices.has(childPubKey)) {
```
should we also check that the child doesn't exist in custodySigners?

pfletcherhill replied:
Oh this brings up an interesting edge case...what happens when the same key is added as a custody signer on one hub and as a delegate on another hub? We don't currently handle that conflict.

pfletcherhill replied:
Okay, upon thinking about this more...
* Custody signers will be identified using addresses (20 bytes) rather than public keys, right? Because we'll look at the `msg.sender` in the account registry contract.
* Delegate signers now are identified using their public keys (32 bytes)
* If we keep this distinction—custody signers are addresses, delegate signers are public keys—we won't have conflicts between the two

pfletcherhill replied:
This PR addresses this https://github.com/farcasterxyz/hub/pull/50

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    this._messages = new Map();
+  }
+
+  private get _vertices() {
```
Fair enough. I was undecided about where to put private getters, but I can move them down.

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+
+/*   
+  SignerSet manages the account's associated custody addresses authorized 
+  for Signing and their corresponding delegates.
```
Added summary at top of set

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    return err('SignerSet.merge: invalid message format');
+  }
+
+  mergeSignerAdd(message: SignerAdd): Result<void, string> {
```
Made these private in favor of just using `merge`

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  private getEdgesByParent(edges: Map<string, string>) {
+    const byChildMap = new Map<string, Set<string>>();
```
Ah sloppy me

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    return byChildMap;
+  }
+
+  private add(parentPubKey: string, childPubKey: string, hash: string): Result<void, string> {
```
I'd prefer to keep it `add`, because I have a `removeEdge` method below that takes an edge key (tuple string) as argument. And the `add` and `remove` methods add and remove both vertices and edges, not just edges.

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+   * Private Methods
+   */
+
+  private getEdgesByChild(edges: Map<string, string>) {
```
Good call, adding comment

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+
+      // If child exists in vAdds
+      else if (this._vertexAdds.has(childPubKey)) {
+        // Get all parents of parentPubKey by traversing edgeAdds
```
Added comment

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    return new Map([...this._edgeAdds, ...this._edgeRemoves]);
+  }
+
+  private get _edgeAddsByChild() {
```
Hmm `_edgeAddsByChild` makes more sense to me, since it's the result of `getEdgesByChild`. And `AllChildren` implies all children in both adds and removes to me.

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+  }
+
+  mergeSignerAdd(message: SignerAdd): Result<void, string> {
+    const parentPubKey = message.signer;
```
Based on our discussion before, this will be the custody's address. So I may rename this to `parentKey` to be slightly less confusing.

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+        { transient: { signer: custodySigner } }
+      );
+      expect(set.merge(remC).isOk()).toBe(false);
+      expect(vRems()).toEqual(new Set());
```
do we expect the edge to be retained here?

pfletcherhill replied:
The edge from `remC`? No, because it's not a valid edge (yet) and the remove message doesn't contain enough info to verify whether it is valid or invalid.

varunsrin commented on src/factories.ts:
```diff
@@ -12,29 +12,55 @@ import {
+  SignatureAlgorithm,
+  Message,
+  MessageFactoryTransientParams,
+  MessageSigner,
```
nit: message signer isn't used here

varunsrin commented on src/utils.ts:
```diff
@@ -106,3 +106,14 @@ export const generateEd25519KeyPair = async (): Promise<KeyPair> => {
+  return { privateKey, signerKey, type: SignatureAlgorithm.Ed25519 };
+};
+
+export const generateEthereumSigner = async (): Promise<EthereumSigner> => {
```
nit: might be worth adding a comment about this being an ECDSA signature that follows EIP-191

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+      expect(set.merge(addA).isOk()).toBe(true);
+      expect(set.merge(remA).isOk()).toBe(true);
+      expect(set.merge(remA).isOk()).toBe(true);
+    });
```
another case: `a child is added by 2 different parents and then deleted by 2 different parents`

addA, addBChildOfA, addCChildofA, addDChildOfB, addDChildOfC, remDChildofB, remDChildOfC

varunsrin replied:
another case: `a remove that is added within a sub-tree after the sub-tree has been removed` 

addA, addBChildOfA, addCChildofB, remBChildOfA, addDChildofC

varunsrin replied:
another case: `a child is removed after its parent has already been removed` 

addA, addBChildOfA, addCChildofB, remBChildOfA, remCChildOfB

pfletcherhill replied:
Added tests for all of these

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+      expect(messages().keys()).not.toContain(addCToA.hash);
+    });
+
+    test('succeeds when child has already been deleted by another parent and moves new edge to edgeRemoves', () => {
```
test case `a child is added into a subtree that was removed`

addA, addBChildOfA, addCChildofB, remA

pfletcherhill replied:
Added test

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+      expect(set.merge(addA).isOk()).toBe(true);
+      expect(set.merge(addCToA).isOk()).toBe(true);
+      expect(set.merge(addB).isOk()).toBe(true);
+      expect(set.merge(remCFromA).isOk()).toBe(true);
```
should we also test that changing the order converges to the same state? or is that implicitly covered elsewhere?

addA -> remCFromA -> addCToA -> addB

pfletcherhill replied:
addA -> remCFromA -> addCToA -> addB will not match addA -> addCToA -> addB -> remCFromA, because c has to exist before it can be removed. That failure is covered in another test (`"fails when child hasn't been added yet"`)

varunsrin reviewed (changes requested):
One concern I have is that the test suite isn't exhaustive because there are so many little edge cases - I'll try to do another pass on this before we merge

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,397 @@
+    return Array.from(this._vertices);
+  }
+
+  getEdgeKey(parentPubKey: string, childPubKey: string): string {
```
Let's default to using lowercased versions—it's easier to convert to lowercase rather than convert to checksum case. I just made that change [here](https://github.com/farcasterxyz/hub/pull/50/commits/99eb20f4f13149a221977f1ef4a44166b17d3ed1) for verifying messages and will make the edge keys lowercased in the signer set as well.

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+  test('succeeds with multiple new custody signers', () => {
+    expect(set.addCustody(custody1).isOk()).toBe(true);
+    expect(set.addCustody(custody2).isOk()).toBe(true);
+    expect(custodySigners()).toContain(custody1);
```
Sure, changing to `toEqual(...)`

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+          expect(set.merge(addA).isOk()).toBe(true);
+          expect(set.merge(addCToA).isOk()).toBe(true);
+          const addCToADuplicateFail: SignerAdd = { ...addCToADuplicate, hash: addCToA.hash.slice(0, -1) };
+          expect(set.merge(addCToADuplicateFail).isOk()).toBe(true);
```
Yeah, good catch

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+          expect(vAdds().size).toEqual(3);
+          const addCToBHigherHash: SignerAdd = { ...addCToB, hash: addCToA.hash + 'a' };
+          expect(set.merge(addCToBHigherHash).isOk()).toBe(true);
+          expect(vAdds().size).toEqual(3);
```
Updated tests to check for specific values of vAdds, vRems, eAdds, eRems, and messages for each test

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+          expect(vAdds().size).toEqual(3);
+          const addCToBLowerHash: SignerAdd = { ...addCToB, hash: addCToA.hash.slice(0, -1) };
+          expect(set.merge(addCToBLowerHash).isOk()).toBe(true);
+          expect(vAdds().size).toEqual(3);
```
Updated

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+    expect(set.addCustody(custody1).isOk()).toBe(true);
+    expect(custodySigners()).toEqual(new Set([custody1]));
+  });
+});
```
Yeah, added

pfletcherhill commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,284 @@
+      expect(messages().keys()).not.toContain(remA.hash);
+    });
+
+    test('succeeds and removes subtree', () => {
```
Updated with additional descendants and a peer B that is not removed

pfletcherhill commented:
Summary of changes since last review:
* Expand signer set test suite to check full state (i.e. `vAdds`, `vRems`, `eAdds`, `eRems`, and `messages`) after every test
* Add tests for all the cases mentioned in @varunsrin's review
* Add explanatory comments to signer set file and methods
* Update cycle prevention logic to only reject adds that would beat out other edges and cause a cycle in eAdds
* Add more cycle tests
* Support removing custody addresses
* Sanitize keys coming from all public methods, so we always use lowercase versions in signer set

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,639 @@
+  set._reset();
+});
+
+describe('addCustody', () => {
```
a case that is currently uncovered by tests: 

- block 1: custody A is added
- block 2: custody B is added
- block 3: custody B removes custody A with a RemoveAll
- block 4: custody A is added again  

this tests line 122 in signerSet `this._custodyRemoves.delete(sanitizedAddress);`

pfletcherhill replied:
Adding

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,639 @@
+    ]);
+  });
+
+  test('succeeds (no-ops) with a duplicate custody address', () => {
```
nit: maybe 'succeeds idempotently with a duplicate custody address' ?

pfletcherhill replied:
Eh I don't know. I always misspell idempotent when I'm typing it out. You definitely think it's better?

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+   *
+   * @param event - event from Farcaster ID Registry
+   */
+  addCustody(event: CustodyAddEvent): Result<void, string> {
```
it is confusing that `addCustody` and `merge` are semantically equivalent operations with different names

why merge in one place vs add<Item> in the other?

pfletcherhill replied:
Updated

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    // If custody exists in custodyRemoves
+    const existingCustodyRemove = this._custodyRemoves.get(sanitizedAddress);
```
nit: should we use `Rem` and `Rems` instead of `Remove` and `Removes` everywhere? brevity might make the code easier to read

pfletcherhill replied:
Eh I'm on the fence about this one. I went with "remove" because the remove types (`CastRemove` `VerificationRemove`, etc) will always have it spelled out, and I thought the sets should be consistent. But I could go either way...

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+
+    // If custody exists in custodyAdds
+    const existingCustodyAdd = this._custodyAdds.get(sanitizedAddress);
+    if (existingCustodyAdd) {
```
nit: it may be slightly clearer if we combined these clauses into a single statement and expanded on the comment. something like: 

```ts
   // If the incoming add operation is an earlier block or the same block as the prior add operation, ignore it
    if (priorCustodyAdd && priorCustodyAdd.blockNumber >= event.blockNumber) {
       return ok(undefined);
    }

```

pfletcherhill replied:
fwiw I think "existing" is better than "prior" because when we define that variable we don't know whether it is prior or not, just that it exists or doesn't

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+      const existingCustodyAddEvent = this._custodyAdds.get(sanitizeSigner(existingCustodyRemove.signer));
+
+      if (existingCustodyAddEvent) {
+        // If existing block number wins (same or greater block), no-op
```
more details here would be helpful to readers

```suggestion
        // If the incoming add operation adds an address that was previously removed
        // and if the address that performed the remove operation is more recent than the incoming add operation
        // do nothing
```

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    if (existingCustodyRemove) {
+      const existingCustodyAddEvent = this._custodyAdds.get(sanitizeSigner(existingCustodyRemove.signer));
+
+      if (existingCustodyAddEvent) {
```
i assume it is impossible to be in the state `existingCustodyRemove && ! existingCustodyAddEvent == true?` because if the addevent was removed then the remove event should also be removed?

if so, should we be throwing? because it seems like we will just ignore this case right now

pfletcherhill replied:
Yeah, should be impossible. I added "unexpected state" errors here and a few other places we have a similar pattern.

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+        // If existing block number wins (same or greater block), no-op
+        if (existingCustodyAddEvent.blockNumber >= event.blockNumber) return ok(undefined);
+
+        // If new block number wins, remove address from removes
```
expanding on this comment in the prior style might also be helpful

varunsrin reviewed (commented):
completed review on `addCustody` method

varunsrin commented:
@pfletcherhill the coverage report here also shows a few untested lines in other functions that are probably worth adding coverage to: https://app.codecov.io/gh/farcasterxyz/hub/compare/38/tree/src/sets/signerSet.ts

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+   */
+
+  /**
+   * mergeCustodyRemoveAll moves all address in custodyAdds with a block number before the
```
nit: `addresses`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+   * Custody addresses that were added in the same block as the signer of the CustodyRemoveAll
+   * message will not be removed.
+   *
+   * @param message - CustodyRemoveAll message
```
nit: we should either remove this or specify more detail like `signed message that indicates that all prior custody addresses should be removed from the valid set`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+   * @param message - CustodyRemoveAll message
+   */
+  private mergeCustodyRemoveAll(message: CustodyRemoveAll): Result<void, string> {
+    const sanitizedAddress = sanitizeSigner(message.signer);
```
nit: should we just call this `lowerCaseAddress` instead? clearer about what it does

varunsrin replied:
similar nit about just calling `message.signer.toLowerCase()` instead of `sanitizeSigner(message.signer)` which is just as verbose but less clear about what its doing

pfletcherhill replied:
Added a comment from the last PR about this: https://github.com/pfletcherhill/farcaster-hub/pull/3#discussion_r942905953

If we decide to change from lowercase to uppercase for some reason, that's trivial to do if we used `sanitizeSigner` everywhere.

varunsrin replied:
ah ok fair, thats a good argument for the abstraction

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+  private mergeCustodyRemoveAll(message: CustodyRemoveAll): Result<void, string> {
+    const sanitizedAddress = sanitizeSigner(message.signer);
+
+    // If custodyAddress exists in custodyRemoves, no-op
```
the comment is so similar to the logic that it doesn't really add much here. 

maybe something like `If the address that requested the remove all been removed by another another remove all message, do nothing`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    // If custodyAddress exists in custodyRemoves, no-op
+    if (this._custodyRemoves.has(sanitizedAddress)) return ok(undefined);
+
+    // If custodyAddress does not exist in custodyAdds, fail
```
`if the add message for this custody address was never received, fail the operation`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    const custodyAddEvent = this._custodyAdds.get(sanitizedAddress);
+    if (!custodyAddEvent) return err('SignerSet.mergeCustodyRemoveAll: custodyAddress does not exist');
+
+    // For each custody address, remove if block number is before signer
```
`Remove each custody address and their child delegate signers where the custody address if the custody address was from a block that preceeds the block in which this signer became a custody address`

varunsrin replied:
add an explicit note about how we don't care about events in the same block

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+        // Before we remove the custody address, we should revoke the signer by dropping all messages signed
+        // by that custody address from the set
+
+        // Drop all SignerAdd messages signed by the custody address being removed
```
add a note about why its a drop and not a move into the remove set

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,639 @@
+        expect(events).toEqual([['addCustody', custody1.signerKey]]);
+      });
+
+      test('succeeds and removes signers added by removed custody address', () => {
```
nit: `succeeds and drops signers`

varunsrin replied:
nit: `succeeds and drops signers`

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,639 @@
+          ['removeCustody', custody1.signerKey],
+        ]);
+      });
+
```
test case:  add signers, remove signers and then remove the custody address and test that the removes were dropped as well

varunsrin commented on src/sets/signerSet.test.ts:
```diff
@@ -0,0 +1,639 @@
+          ['removeCustody', custody1.signerKey],
+        ]);
+      });
+
```
test case:  add custody address A, do a rem all from A, then add B, do a rem all from B and assert that the rem all from A was dropped

pfletcherhill replied:
Ah this is a bug! Good catch! We should update the past remove messages to reference a later block number rather than dropping them.

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    const custodyAddress = sanitizeSigner(message.signer);
+    const signerKey = sanitizeSigner(message.data.body.delegate);
+
+    // If custody address has been removed, no-op
```
`If the custody address that authorized the signer has been removed by another operation, do nothing`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    // If custody address has been removed, no-op
+    if (this._custodyRemoves.has(custodyAddress)) return ok(undefined);
+
+    // If custody address is missing, fail
```
`If the custody address that authorized the signer was never added, do nothing`

varunsrin commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+        // If block numbers are the same
+        if (existingCustodyAddEvent.blockNumber === custodyAddEvent.blockNumber) {
+          // If the custody addresses are the same and existing signer add has the same or higher hash, no-op
+          if (
```
can we do this more deterministcally by requiring a transactionId and logIndex to be passed in along with the block? 

that way we can determine exactly which custodyAdd event wes later and choose accordingly

pfletcherhill commented on src/sets/signerSet.ts:
```diff
@@ -0,0 +1,364 @@
+    return this._signerAdds.get(signer);
+  }
+
+  merge(message: SignerMessage | CustodyRemoveAll): Result<void, string> {
```
See comment here https://github.com/farcasterxyz/hub/pull/38/#discussion_r944584009
